modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1700,19 +1700,33 @@ void GDScriptAnalyzer::resolve_for(GDScriptParser::ForNode *p_for) {
 								reduced = Vector3i(args[0], args[1], args[2]);
 								break;
 						}
+
 						p_for->list->is_constant = true;
 						p_for->list->reduced_value = reduced;
-					}
-				}
+						p_for->list->set_datatype(type_from_variant(reduced, p_for->list));
+					} else {
+						Variant::Type range_replacement = Variant::NIL;
+						switch (call->arguments.size()) {
+							case 1:
+								range_replacement = Variant::INT;
+								break;
+							case 2:
+								range_replacement = Variant::VECTOR2I;
+								break;
+							case 3:
+								range_replacement = Variant::VECTOR3I;
+								break;
+						}
```

Comment:
One issue here is that technically floats are allowed, but this is forcing it to be int in all cases. This is tricky because there's no way to know until runtime. I wonder if this needs a special instruction in the VM so it can properly check the types (also would allow to have a more specific error message, as you asked).

Reply:
I think we need `rangef` or `range_float` since `for` already supports `float`, `Vector2`, `Vector3`.